### PR TITLE
Reconcile RoleBinding and ClusterRoleBinding resources 

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
Reconcile RoleBinding and ClusterRoleBinding on Subjects and RoleRefs attributes.

Also includes unit tests for RBAC's reconciliation.

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:37]: This function name needs to be changed. as its part of `Reconciler` it reads as if any object can be passed to it and will be created or updated.

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:45]: not sure I understand this variable. `avoidIfServiceAccount` service account is a resource type, but this var is passed in as true for the resource type RoleBinding and Role?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:75]: if we are going to do this, lets just do...
`err = r.getUpdateFunction(existingCopy)()`

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:238]: this function is almost 100 lines. Let's break break it apart.

Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/rbac.go:238]: the verb `change` and `changed` also reads weird. is this modified, or updated? By the name I can't tell if this is a compare function and its telling me something is different or if it's an action function and something was updated.

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:280]: are we guaranteed order?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:311]:
```
changed = changeExistingSubjectsByRequired(&existingRoleBinding.Subjects, &requiredRoleBinding.Subjects) || changeExistingRoleRefByRequired(&existingRoleBinding.RoleRef, &requiredRoleBinding.RoleRef)
```

or 

```
subjectsDifferent := changeExistingSubjectsByRequired(&existingRoleBinding.Subjects, &requiredRoleBinding.Subjects)
// short circuit if the first is true. 
changed = subjectsDifferent || changeExistingRoleRefByRequired(&existingRoleBinding.RoleRef, &requiredRoleBinding.RoleRef) 
```

also applies to the ClusterRoleBinding case

**Review by ashleyschuett — CHANGES_REQUESTED** [review]
Some code comments. 

At a higher level was was discovered about the different types and have we tested in a running cluster? Which fields are defaulted/added that could break this logic for each type and each subject/rules/role ref type?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:37]: There are 4 dedicated functions to reconcile rbac objects which are `createOrUpdateClusterRole`, `createOrUpdateClusterRoleBinding`, `createOrUpdateRole` and `createOrUpdateRoleBinding`. This is meant to serve as a private helper function.

Thanks, you're right, it is confusing.
I think that I will change this function from being a "method" into being a regular private function to reduce the confusion. Will also rename the function to something like `rbacObjectUpdateHelper` so that its purpose will be clear.

What do you think?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:45]: I saw that we already began to [discussed this](https://github.com/kubevirt/kubevirt/pull/5224#discussion_r600294351) in my last reconciliation [PR](https://github.com/kubevirt/kubevirt/pull/5224).

The following condition to skip reconciliation of `"kubevirt-monitoring"` resources existed for `Role` and `RoleBinding` before I made any changes to rbac:
```
if !r.stores.ServiceMonitorEnabled && (rb.Name == rbac.MONITOR_SERVICEACCOUNT_NAME) {
	return nil
}
```
Do you think there's a reason to remove this condition?

Anyways, I understand your confusion. I think one plausible solution is to add a private `toSkipReconciliation(obj)` function. There any conditions to skip the reconciliation can be defined. What do you think?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:238]: Sure, `modified` is clearer.

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:280]: Good point. Thanks.
I will also add a test for trying to make no changes but changing the order of subjects.

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:311]: The whole idea here is to avoid short circuit if the first is true.

Please note that `changeExistingSubjectsByRequired` and `changeExistingRoleRefByRequired` functions update the "existing" object by the required one. So if `changeExistingSubjectsByRequired` returned `true` it doesn't mean that I don't want to execute `changeExistingRoleRefByRequired` - I want to always execute both functions in order the perform all necessary updates. Only at the end I need to know if any modification occured.

But again, I see where the confusion is coming from.
I think it's best to break it into two variables: `subjectsUpdated` and `roleRefUpdated`, or something similar.

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/rbac.go:45]: I don't think that's needed. I think you can remove this check and instead add it before line  30 && 34. We call this function for each one individually and we do nothing if its true so I'd check before calling the function and remove the func param.

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/rbac.go:311]: Do we need to update the existing copy and update using it? could we instead just check if something is different and if it is update to the required spec? then we could short circuit as soon as something doesn't match

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:238]: Regarding breaking the function apart, the reason this function is long is because helper functions are defined within it (`changeExistingPolicyRulesByRequired`, `changeExistingSubjectsByRequired`, ...). These helper functions are there, as stated in the comment, for performance reasons since it's faster than a simple deep reflection comparison.

IMHO if a long function is organized in a readable way and defines helper functions for its context that's not a problem per se. I think the alternative, which is extracting out these helper functions to regular private functions, is less clear since these helper functions belong to a very specific context which is now gone. In the future we could end up having a lot more functions in `rbac.go` and these would not be obvious and would make a mess.

Does that make sense to you?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:311]: I saw that this is how it's implemented in OpenShift, there they update `existing` by whatever necessary from `required`. My guess is that they did it this way to be on the safe side (not taking anything from `required` implicitly).

But you're right, there is a performance overhead for this approach. Not sure it's worth it.

What do you think?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [inline pkg/virt-operator/resource/apply/rbac.go:283]: instead of a nested loop which has N2 complexity I would sort and then compare, but the list should also be relatively short so I guess it doesn't really matter :shrug:

**Review by ashleyschuett — APPROVED** [review]
seems okay.. can you confirm that you have tested on a running cluster and have made changes to the different types and that there is no hot looping?

**Comment by ashleyschuett** [comment]
/hold

would like confirmation before merge.

**Comment by iholder-redhat** [comment]
> seems okay.. can you confirm that you have tested on a running cluster and have made changes to the different types and that there is no hot looping?

Yes, I confirm I've tested it on a running local cluster.
I've changed values of different RoleBindings / ClusterRoleBindings and verified that are being synced back to the original value. I've also verified that the operator logs don't include more than one update for each one that I made.

Thanks @ashleyschuett for all your feedback and review!

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:283]: Exactly what I thought.
IMHO, when performance overhead is negligible (i.e. if we have very small arrays that are not supposed to ever grow in a significant way), it's better to keep the code simple.

File: pkg/virt-operator/resource/apply/rbac_test.go
Comment (enp0s3) [inline pkg/virt-operator/resource/apply/rbac_test.go:1]: @iholder-redhat Can you please add a GPL header?

**Review by enp0s3 — CHANGES_REQUESTED** [review]
@iholder-redhat Hi. A small comment below, besides that looks good

File: pkg/virt-operator/resource/apply/rbac_test.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac_test.go:20]: Done. Thanks for noticing!

File: pkg/virt-operator/resource/apply/rbac.go
Comment (dhiller) [inline pkg/virt-operator/resource/apply/rbac.go:240]: nit: I understand that the naming is throughout the code like this, i.e. `existing` vs. `required`, I find the wording still a bit confusing as what we are doing here is reconcile the current to a desired state, which is why I would name the method 
```
reconcileRBAC(current, desired runtime.Object) (modified bool) {
```

**Review by dhiller — APPROVED** [review]
/approve
/hold

Minor readability nits, otherwise looks great! @rmohr any objections?

Side note: I still don't get it that such a fundamental thing like comparison does not have a standard implementation.... Or am I missing something?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:45]: I see your point. My intention was that it's a `CreateOrUpdate` function that's private to `rbac`.
Maybe `createOrUpdateHelper` is better?

File: pkg/virt-operator/resource/apply/rbac.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:240]: I agree, your terminology is better. Thanks!

**Comment by iholder-redhat** [comment]
> Side note: I still don't get it that such a fundamental thing like comparison does not have a standard implementation.... Or am I missing something?

@dhiller The thing is, IIUC, that in golang the only two ways to compare between 2 slices is either by performing `reflect.DeepEqual`, which is expensive in terms of performance, or implementing it by hand (compare length, each element, etc').

_Note: IIUC, you can compare two **arrays**, meaning slices with a fixed length (`[2]string{"a", "b"}`), but this is not the case for us. See this for more info: https://yourbasic.org/golang/compare-slices/._

In one of my previous reconciliation PRs I had a comment that maybe it'd be better to have a private function that compares slices then use it everywhere to reduce this overhead. This might be a bit of an overkill. WDYT?

**Comment by dhiller** [comment]
> > Side note: I still don't get it that such a fundamental thing like comparison does not have a standard implementation.... Or am I missing something?
> 
> @dhiller The thing is, IIUC, that in golang the only two ways to compare between 2 slices is either by performing `reflect.DeepEqual`, which is expensive in terms of performance, or implementing it by hand (compare length, each element, etc').
> 
> _Note: IIUC, you can compare two **arrays**, meaning slices with a fixed length (e.g. `[2]string{"a", "b"}`), but this is not the case for us. See this for more info: https://yourbasic.org/golang/compare-slices/._
> 
> In one of my previous reconciliation PRs I had a comment that maybe it'd be better to have a private function that compares slices then use it everywhere to reduce this overhead. This might be a bit of an overkill. WDYT?

@iholder-redhat to be clear, this was not targeted at you, as it clearly is the recommendation to do it like you did. I just find it annoying that everyone has to reinvent the wheel and produce duplicate code. But there must be a reason for this, even though I am missing it. </nagging> ;)

**Comment by iholder-redhat** [comment]
> > > Side note: I still don't get it that such a fundamental thing like comparison does not have a standard implementation.... Or am I missing something?
> > 
> > 
> > @dhiller The thing is, IIUC, that in golang the only two ways to compare between 2 slices is either by performing `reflect.DeepEqual`, which is expensive in terms of performance, or implementing it by hand (compare length, each element, etc').
> > _Note: IIUC, you can compare two **arrays**, meaning slices with a fixed length (e.g. `[2]string{"a", "b"}`), but this is not the case for us. See this for more info: https://yourbasic.org/golang/compare-slices/._
> > In one of my previous reconciliation PRs I had a comment that maybe it'd be better to have a private function that compares slices then use it everywhere to reduce this overhead. This might be a bit of an overkill. WDYT?
> 
> @iholder-redhat to be clear, this was not targeted at you, as it clearly is the recommendation to do it like you did. I just find it annoying that everyone has to reinvent the wheel and produce duplicate code. But there must be a reason for this, even though I am missing it. `</nagging>` ;)

Sure, I totally see your point, this is weird that there is no acceptable solution for this.. Maybe it's just because golang is pretty new? (still waiting for templates... ;) ).

In any case, maybe it's better to have a `comparison_util.go` file somewhere and then we can at least implement it once for all.

And this is not nagging, thanks for raising this discussion :)

File: pkg/virt-operator/resource/apply/rbac.go
Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/rbac.go:45]: `createOrUpdateHelper` is confusing because once again it sounds like it can create or update any object since the param type is `runtime.Object`

Comment (iholder-redhat) [reply pkg/virt-operator/resource/apply/rbac.go:45]: `createOrUpdateRbacObject`?

**Comment by dhiller** [comment]
/unhold

I think that my comments are nits, and that they should not block this to get merged.

**Pull request merged**
